libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2078,6 +2086,112 @@ void TypeChecker::typeCheckABIEncodeFunctions(
 	}
 }
 
+void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCall)
+{
+	vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
+
+	// Expecting first argument to be the function pointer and second to be a tuple.
+	if (arguments.size() != 2)
+	{
+		m_errorReporter.typeError(
+			6219_error,
+			_functionCall.location(),
+			"Expected two arguments: a function pointer followed by a tuple."
+		);
+		return;
+	}
+
+	auto const functionPointerType = dynamic_cast<FunctionTypePointer>(type(*arguments.front()));
+
+	if (!functionPointerType)
+	{
+		m_errorReporter.typeError(
+			5511_error,
+			arguments.front()->location(),
+			"Expected first argument to be a function pointer, not \"" +
+			type(*arguments.front())->canonicalName() +
+			"\"."
+		);
+		return;
+	}
+
+	if (functionPointerType->kind() != FunctionType::Kind::External)
+	{
+		SecondarySourceLocation ssl{};
+
+		if (functionPointerType->hasDeclaration())
+			ssl.append("Function is declared here:", functionPointerType->declaration().location());
+		m_errorReporter.typeError(
+			3509_error,
+			arguments[0]->location(),
+			ssl,
+			"Function must be \"public\" or \"external\". Did you forget to precede \"this.\"?"
+		);
+	}
+
+	// There are no checks we can do if all types are accepted
+	if (functionPointerType->takesArbitraryParameters())
+	{
+		SecondarySourceLocation ssl{};
+
+		if (functionPointerType->hasDeclaration())
+			ssl.append("Function is declared here:", functionPointerType->declaration().location());
+		m_errorReporter.typeError(
+			6716_error,
+			arguments[0]->location(),
+			ssl,
+			"Function takes arbitrary parameters. No type-checking possible."
+		);
+		return;
+	}
+
+	// Tuples with only one component become that component
+	auto const argumentTuple = std::dynamic_pointer_cast<TupleExpression const>(arguments[1]);
+	size_t const numArguments = argumentTuple ? argumentTuple->components().size() : 1;
+	size_t const numParameters = min(numArguments, functionPointerType->parameterTypes().size());
+
+	if (functionPointerType->parameterTypes().size() != numArguments)
+		m_errorReporter.typeError(
+			7788_error,
+			_functionCall.location(),
+			"Expected " +
+			to_string(functionPointerType->parameterTypes().size()) +
+			" instead of " +
+			to_string(numArguments) +
+			" components for the tuple parameter."
+		);
+
+	// Get i'th component if it's a tuple or just the argument itself if it's not.
+	auto getArg = [&](size_t _i) -> ASTPointer<Expression const>
+	{
+		solAssert(_i < numParameters, "");
+
+		if (!argumentTuple)
+			return arguments[1];
+
+		return argumentTuple->components()[_i];
```

Comment:
Maybe this is a bit simpler?
```suggestion
		if (argumentTuple)
			return argumentTuple->components().at(_i);
		else
			return arguments.at(1);
```

Reply:
If you want to add an assertion, then what I would suggest is adding `solAssert(_i == 0)` in the case where we do not have an argument tuple.